### PR TITLE
Core code generation

### LINKED ISSUE #3: Compile s-expressions into bytecode

### PR DESCRIPTION
**Pull request by james-perretta**
- Emits bytecode for all language constructs (does not include definitions of builtin functions or the array bounds/type checking to go with them)
- Adds SCall, SQuote, SBool, and BoolLit types
- Removes Environment class in favor of dictionaries (since we only have local and global environments)

Fixes #3

### CODE REVIEW

**Comment by porglezomp** [comment]
Woo! Big review! I'll see how much of this I can get to tonight, but I'm also working on OS stuff.

File: bytecode.py [hunk 1]
```diff
@@ -67,20 +66,31 @@ def __str__(self) -> str:
...
+        return self.value
+
+    def __str__(self) -> str:
+        return f"'{self.value}"
+
+
 class EvalEnv:
     _local_env: Dict[Var, Value]
-    _global_env: Environment
+    _global_env: Dict[SSym, SExp]
```
Comment (porglezomp) [inline bytecode.py]: Should this maybe be `Dict[SSym, Value]`?

File: lint.sh [hunk 2]
```diff
@@ -4,5 +4,5 @@ status=0
 mypy --strict ./*.py || status=1
 pycodestyle ./*.py || status=1
 # for some reason Travis-CI isort doesn't know about dataclasses???
-isort --diff --check-only --recursive -b dataclasses || status=1
+isort --diff --check-only --recursive -b dataclasses *.py || status=1
```
Comment (porglezomp) [inline lint.sh]: Doesn't `--recursive` find all relevant files automatically? I get warnings as I would expect without having to include this.

File: scheme.py [hunk 3]
```diff
@@ -132,6 +129,15 @@ def __str__(self) -> str:
         return f"({str(self.first)} . {str(self.second)})"
 
 
+@dataclass(frozen=True)
+class Quote(SExp):
+    """A quoted list"""
+    slist: SList
```
Comment (porglezomp) [inline scheme.py]: Quotes are arbitrary s-expressions, not just lists. The most important non-list case is being able to write `'symbol`.

File: scheme.py [hunk 4]
```diff
@@ -208,7 +214,13 @@ def type_name(self) -> SSym:
         return SSym('function')
 
 
-@dataclass
+@dataclass(frozen=True)
+class SCall(SExp):
+    func: Union[SSym, SFunction]
```
Comment (porglezomp) [inline scheme.py]: This doesn't allow for cases like the following:
```lisp
((if p fun1 fun2) args)
```
`func` should be an arbitrary s-expression.

File: lint.sh (hunk 2 above)
Comment (james-perretta) [reply lint.sh]: Ah, so I have my virtual environment folder in the same directory as the py files, so I was getting a ton of errors in library files. I didn't mean to commit the change though and can undo it.

File: bytecode.py (hunk 1 above)
Comment (james-perretta) [reply bytecode.py]: Sure

File: scheme.py (hunk 3 above)
Comment (james-perretta) [reply scheme.py]: Can you give an example of a program we want to be able to write that needs a standalone symbol literal?

Comment (james-perretta) [reply scheme.py]: Nevermind, I'll update it

File: scheme.py (hunk 4 above)
Comment (james-perretta) [reply scheme.py]: Updated and added a test case

File: emit_IR.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+
+def name_generator(prefix: str) -> Iterator[str]:
+    count = 0
+    while True:
+        yield f'{prefix}{count}'
+        count += 1
+
+
+class EnvBindingNotFound(Exception):
```
Comment (porglezomp) [inline emit_IR.py]: When I was reading custom exceptions stuff, I found things suggesting that you want all of your exceptions in one file, which is why we have the `errors.py`. I *think* this should go there?

**Review by porglezomp — CHANGES_REQUESTED** [review]

File: emit_IR.py (hunk 5 above)
Comment (james-perretta) [reply emit_IR.py]: I'm familiar with two philosophies: having all the exceptions in one place, and having exceptions defined in the same module that they logically "originate" from. If we want to go with all the exceptions in errors.py that's fine with me. 

We should probably also have a custom base class for our custom exceptions, no?

**Review by porglezomp — APPROVED** [review]

**Pull request merged**